### PR TITLE
[feg] initial commit for s8_proxy

### PR DESCRIPTION
**Pull request by uri200**
Signed-off-by: Oriol Batalla <obatalla@fb.com>

## Summary
This is the initial PR that includes the new s8_proxy. 

The intention of this PR is to build a running service with just one call (Create Session Request). Neither the other calls or their protos are still defined.

We have also created a wrapper of the gtp client provided by [go-gtp](https://github.com/wmnsk/go-gtp). That wrapper includes some functions to ease the creation. It also adds a layer of abstraction that will be used on future PRs

## Test Plan

There is two tests, 
- GTP client test: on `gtp_client_test.go` that tests echo request on a basic pair of gtp services.
- S8_prox test: on `s8_proxy_test.go` that runs a create session request. It uses a mock PGW on`s8_proxy/servicers/mock_pgw` implements `CreateSessionResponse` (for now)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: feg/gateway/gtp/gw.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
```
Comment (uri200) [inline feg/gateway/gtp/gw.go:13]: those functions are not used yet

File: feg/gateway/services/s8_proxy/servicers/mock_pgw/cs.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
+/*
```
Comment (uri200) [inline feg/gateway/services/s8_proxy/servicers/mock_pgw/cs.go:1]: maybe i will move this mock_pgw *folder* somewhere else (maybe to gtp folder) or use a different name to indicate it is for test

**Review by themarwhal — APPROVED** [review]

File: feg/gateway/gtp/go.mod [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+go 1.13
+
+replace (
+	github.com/golang/protobuf  => 	github.com/golang/protobuf v1.4.3
+)
+
+require (
+	github.com/wmnsk/go-gtp v0.7.15
+)
```
Comment (emakeev) [inline feg/gateway/gtp/go.mod]: I am not sure, you need this go.mod. Did you try to just include github.com/wmnsk/go-gtp v0.7.15 in feg/gateway/go.mod require block?
I don't see any direct usages of protobuf here & go-gtp module dependencies should be satisfied automatically.

File: feg/gateway/gtp/gtp_client.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gtp
```
Comment (emakeev) [inline feg/gateway/gtp/gtp_client.go:14]: // package gtp ... - package description is missing

File: feg/gateway/gtp/gtp_client.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	*gtpv2.Conn
+	connType   uint8
+	localAddr  *net.UDPAddr
+	remoteAddr *net.UDPAddr
+}
+
+// NewConnectedAutoClient creates a GTP-C client finding out automatically the local Address that needs to
+// be used to reach the remote IP. It checks remote address using echo. And It also runs the GTP-C
+// server waiting for incoming calls.
+func NewConnectedAutoClient(ctx context.Context, remoteAddrStr string, connType uint8) (*Client, error) {
```
Comment (emakeev) [inline feg/gateway/gtp/gtp_client.go:39]: is it IP or address? either comment or function definition need to be changed

File: feg/gateway/gtp/gtp_client.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+	remoteAddr, err := net.ResolveUDPAddr("udp", remoteAddrStr)
+	if err != nil {
+		return nil, fmt.Errorf("could not resolve remote address: %s", err)
```
Comment (emakeev) [inline feg/gateway/gtp/gtp_client.go:42]: fmt.Errorf("failed to resolve remote address '%s': %v", remoteAddrStr , err)

could be more descriptive

File: feg/gateway/gtp/gtp_client.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+func (c *Client) GetLocalAddress() *net.UDPAddr {
+	return c.localAddr
+}
+
+// Get preferred outbound ip of this machine
+func GetOutboundIP(testIp *net.UDPAddr) (net.IP, error) {
+	//Conn, err := net.Dial("udp", "8.8.8.8:80")
+	connection, err := net.Dial("udp", testIp.String())
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't local io running in this server: %s", err)
```
Comment (emakeev) [inline feg/gateway/gtp/gtp_client.go:117]: typo?

**Pull request merged**